[LS] overhauled rename service

JanKoehnlein opened the pull request:
- delegate to rename language of the declaration of the renamed target
- allow to trigger rename with caret right after the identifier
- support versioned documents

Closes #917, closes #916 and closes #1072

dhuebner commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService.xtend:
```diff
@@ -9,59 +9,86 @@ package org.eclipse.xtext.ide.server.rename
+		rename(workspaceManager, renameParams, new Options(false), cancelIndicator)
+	}
+	
+	override rename(WorkspaceManager workspaceManager, RenameParams renameParams, Options options, CancelIndicator cancelIndicator) {
```
I'm not sure if need the @since tag here too

JanKoehnlein replied:
Fixed

cdietrich commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend:
```diff
@@ -607,14 +608,19 @@ import org.eclipse.xtext.findReferences.IReferenceFinder.IResourceAccess
-			renameService.rename(workspaceManager, params, cancelIndicator)
+			if (renameService instanceof IRenameServiceExtension) {
+				val options = new IRenameServiceExtension.Options(params?.capabilities?.workspace?.workspaceEdit?.documentChanges === Boolean.TRUE)
+				renameService.rename(workspaceManager, renameParams, options, cancelIndicator)
```
i still wonder if we can get rid of workspace manager here. e.g. codeaction does not have access to it too.
i also wonder if and how we should maintain api here since lsp/lsp4j changes theirs for each release

cdietrich replied:
@szarnekow @kthoms what do you think?

szarnekow replied:
Iff there was a customized IRenameService, it would no longer be called?

JanKoehnlein replied:
@cdietrich 

@szarnekow why? The LS still injects the old interface and only calls the new method if it's present. Did I miss something?

szarnekow replied:
My feeling is that you missed that the old impl implements the original interface but no framework code will call it anymore.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter.xtend:
```diff
@@ -31,21 +35,35 @@ class ChangeConverter implements IAcceptor<IEmfResourceChange> {
+		@Inject protected IResourceServiceProvider.Registry registry
 
+		/**
+		 * @deprecated use {@link #create(WorkspaceManager, WorkspaceEdit, boolean}�instead.
```
Looks like the link points to a method that does not exist

JanKoehnlein replied:
Fixed

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter.xtend:
```diff
@@ -31,21 +35,35 @@ class ChangeConverter implements IAcceptor<IEmfResourceChange> {
+	val IRenameServiceExtension.Options options
 	
-	protected new(WorkspaceManager workspaceManager, IResourceServiceProvider.Registry registry, WorkspaceEdit edit) {
+	protected new(WorkspaceManager workspaceManager, IResourceServiceProvider.Registry registry, WorkspaceEdit edit, IRenameServiceExtension.Options options) {
```
Could it be that clients may have specialized the ChangeConverter and the constructor still would need to callable as is.

JanKoehnlein replied:
Yes, that's where I stopped to be API compatible cause I think this is highly unlikely.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter.xtend:
```diff
@@ -31,21 +35,35 @@ class ChangeConverter implements IAcceptor<IEmfResourceChange> {
 		this.registry = registry
 		this.edit = edit
+		this.options = options
+		if (options?.clientSupportsVerisonedDocuments)
```
typo `clientSupportsVerisonedDocuments `

JanKoehnlein replied:
Fixed

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter.xtend:
```diff
@@ -88,12 +106,24 @@ class ChangeConverter implements IAcceptor<IEmfResourceChange> {
 	
-	protected def addTextEdit(URI uri, TextEdit... textEdit) {
-		edit.changes.put(uri.toString, textEdit)
+	protected def addTextEdit(URI theUri, Document document, TextEdit... textEdits) {
```
Again: is subclassing the ChangeConverter a use case?

JanKoehnlein replied:
Again: I don't think so.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService.xtend:
```diff
@@ -9,59 +9,86 @@ package org.eclipse.xtext.ide.server.rename
+	 * @deprecated use {@link #rename(WorkspaceManager, RenameParams, Options, CancelIndicator)} 
+	 *    instead.
+	 */
+	@Deprecated
```
Hmm since this method is no longer called, marking it deprecated appears to be not useful.

Maybe there's a better solution compared to the extension interface? Probably default methods are an option.

Alternatively we explicitly do not care about backwards compatibility and do deliberate binary incompatible changes.

JanKoehnlein replied:
It is called when the `IRenameServiceExtension` hasn't been implemented.

szarnekow replied:
Specializations are likely based on the default impl and its API. Since that one implements the extension interface, none of the old methods on it are a) called or b) available anymore.

JanKoehnlein replied:
Got it. Good catch.

JanKoehnlein commented:
Thanks for the reviews. I have addressed all your comments and I am going to merge this now. 

For future API compatibility, I propose to add `option` parameters to all services and declare only these service interfaces API. It is going to be much easier to just evolve the `option`'s types API compatibly when necessary. We could also use them parameter to pass things like the `WorkspaceManager` when it's needed. We could discuss that here or on Gitter.
